server/server/src/billing/app_store_client.rs

```diff
@@ -0,0 +1,124 @@
+) -> Result<RequestBuilder, ServerError<UpgradeAccountAppStoreError>> {
+    let mut header = Header::new(Algorithm::ES256);
+    header.kid = Some(config.iap_key_id.clone());
+
+    let iat = (get_time().0 / 1000) as usize;
+    let exp = (get_time().0 / 1000 + 1200) as usize;
+
+    let claims = Claims {
+        iss: config.issuer_id.to_string(),
+        iat,
+        exp,
+        aud: AUDIENCE.to_string(),
+        bid: BUNDLE_ID.to_string(),
+    };
+
+    let token = encode(&header, &claims, &EncodingKey::from_ec_pem(config.iap_key.as_bytes())?)?;
+    Ok(request
+        .header("alg", ALG)
+        .header("kid", &config.iap_key_id)
+        .header("typ", TYP)
+        .bearer_auth(token))
+}
+
+pub async fn get_sub_status(
+    client: &reqwest::Client, config: &AppleConfig, original_transaction_id: &str,
+) -> Result<(LastTransactionItem, TransactionInfo), ServerError<UpgradeAccountAppStoreError>> {
+    let resp = gen_auth_req(
+        config,
+        client.get(format!("{}/{}", config.sub_statuses_url, original_transaction_id)),
+    )?
+    .send()
+    .await?;
+
+    let resp_status = resp.status().as_u16();
+    match resp_status {
+        200 => {
+            debug!("Successfully retrieved subscription status");
+
+            let sub_status: SubsStatusesResponse = resp.json().await?;
+
+            for sub_group in &sub_status.data {
+                if sub_group.sub_group == config.monthly_sub_group_id {
+                    let last_trans = sub_group
+                        .last_transactions
+                        .get(0)
+                        .ok_or(ClientError(UpgradeAccountAppStoreError::InvalidAuthDetails))?;
+
+                    let part = <&str>::clone(
+                        last_trans
+                            .signed_transaction_info
+                            .split('.')
+                            .collect_vec()
+                            .get(1)
+                            .ok_or_else::<ServerError<UpgradeAccountAppStoreError>, _>(|| {
+                                internal!(
+                                    "There should be a payload in apple jwt: {:?}",
+                                    sub_status
+                                )
+                            })?,
+                    );
+
+                    let trans_info =
+                        serde_json::from_slice(&base64::decode(part).map_err::<ServerError<
+                            UpgradeAccountAppStoreError,
+                        >, _>(
+                            |err| {
+                                internal!(
+                                    "Cannot decode apple jwt payload: {:?}, err: {:?}",
+                                    part,
+                                    err
+                                )
+                            },
+                        )?)?;
+                    return Ok((last_trans.clone(), trans_info));
+                }
```

Comment:
safe to ignore the rest of this data in the loop?

Reply:
Yes it is, we are only looking for a particular subscription. There should not be other things in this loop regardless though.